pkg/operator/ceph/cluster/watcher.go

```diff
@@ -130,6 +144,319 @@ func (c *clientCluster) onK8sNode(ctx context.Context, object runtime.Object) bo
+	if err != nil {
+		return fmt.Errorf("failed to fence cephfs volumes. %v", err)
+	}
+
+	return nil
+}
+
+func getVolumeInUse(cluster *cephv1.CephCluster, volumesInUse []v1.UniqueVolumeName) ([]string, []string) {
+	var rbdVolumesInUse, cephFSVolumeInUse []string
+
+	for _, volume := range volumesInUse {
+		volumesInuseRemoveK8sPrefix := strings.TrimPrefix(string(volume), "kubernetes.io/csi/")
+		splitVolumeInUseBased := strings.Split(volumesInuseRemoveK8sPrefix, "^")
+		logger.Infof("volumeInUse after spliting based on '^' %v", splitVolumeInUseBased)
+
+		if len(splitVolumeInUseBased) == 2 && splitVolumeInUseBased[0] == fmt.Sprintf("%s.rbd.csi.ceph.com", cluster.Namespace) {
+			rbdVolumesInUse = append(rbdVolumesInUse, splitVolumeInUseBased[1])
+
+		} else if len(splitVolumeInUseBased) == 2 && splitVolumeInUseBased[0] == fmt.Sprintf("%s.cephfs.csi.ceph.com", cluster.Namespace) {
+			cephFSVolumeInUse = append(cephFSVolumeInUse, splitVolumeInUseBased[1])
+		}
+	}
+	return rbdVolumesInUse, cephFSVolumeInUse
+}
+
+func listRbdCephFSPV(listPVs *corev1.PersistentVolumeList, cluster *cephv1.CephCluster, rbdVolumesInUse, cephFSVolumeInUse []string) ([]corev1.PersistentVolume, []corev1.PersistentVolume) {
+	var listRbdPV, listCephFSPV []corev1.PersistentVolume
+
+	for _, pv := range listPVs.Items {
+		if pv.Spec.CSI.Driver == fmt.Sprintf("%s.rbd.csi.ceph.com", cluster.Namespace) {
+			for _, rbdVolume := range rbdVolumesInUse {
+				if pv.Spec.CSI.VolumeHandle == rbdVolume {
+					listRbdPV = append(listRbdPV, pv)
+				}
+			}
+		} else if pv.Spec.CSI.Driver == fmt.Sprintf("%s.cephfs.csi.ceph.com", cluster.Namespace) {
+			for _, cephFSVolume := range cephFSVolumeInUse {
+				if pv.Spec.CSI.VolumeHandle == cephFSVolume {
+					listCephFSPV = append(listCephFSPV, pv)
+				}
+			}
+		}
+	}
+	return listRbdPV, listCephFSPV
+}
+
+func (c *clientCluster) fenceRbdVolumes(
+	ctx context.Context, node corev1.Node, cluster *cephv1.CephCluster,
+	clusterInfo *cephclient.ClusterInfo, listRbdPV []corev1.PersistentVolume) error {
+
+	for _, pv := range listRbdPV {
+		logger.Infof("rbd PV NAME %v", pv.Spec.CSI.VolumeAttributes)
+		args := []string{"status", fmt.Sprintf("%s/%s", pv.Spec.CSI.VolumeAttributes["pool"], pv.Spec.CSI.VolumeAttributes["imageName"])}
+		cmd := cephclient.NewRBDCommand(c.context, clusterInfo, args)
+		cmd.JsonOutput = true
+
+		buf, err := cmd.Run()
+		if err != nil {
+			return fmt.Errorf("failed to list watchers for pool/imageName %s/%s. %v", pv.Spec.CSI.VolumeAttributes["pool"], pv.Spec.CSI.VolumeAttributes["imageName"], err)
+		}
+
+		logger.Infof("rbd status to get the client ips %v", string(buf))
+		ips, err := rbdStatusUnMarshal(buf)
+		if err != nil {
+			return fmt.Errorf("failed to unmarshal rbd status output. %v", err)
+		}
+		err = c.createNetworkFence(ctx, pv, node, cluster, ips)
+		if err != nil {
+			return fmt.Errorf("failed to create network fence for node %q. %v", node.Name, err)
+		}
```

Comment:
we need to create NetworkCR only once not for all the PV

Reply:
what if we have more than one volumeInUse in that node, in this case we'll have different client ip's right?

Reply:
cleared below